Aplib.Core/Desire/DesireSet.cs

```diff
@@ -12,22 +13,85 @@ public class DesireSet<TBeliefSet> : IDesireSet<TBeliefSet>
         /// </summary>
         private readonly IGoalStructure<TBeliefSet> _mainGoal;
 
-        /// <inheritdoc />
-        public CompletionStatus Status => _mainGoal.Status;
+        /// <summary>
+        /// Stores the side goal structures of the agent.
+        /// Each of these goal structures has an accompanying guard that must be fulfilled to
+        /// activate the goal structure (i.e., push the goal structure on top of the goal structure stack).
+        /// All active goal structures of the agent that still need to be finished are pushed on the stack.
+        /// </summary>
+        private readonly OptimizedActivationStack
+            <(IGoalStructure<TBeliefSet> goalStructure, System.Func<TBeliefSet, bool> guard)> _goalStructureStack;
+
+        /// <summary>
+        /// If there are no goal structures left to be completed, the status of this desire set is set to the main goal status.
+        /// </summary>
+        public CompletionStatus Status
+            => _goalStructureStack.Count == 0 ? _mainGoal.Status : CompletionStatus.Unfinished;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="DesireSet{TBeliefSet}" /> class.
         /// </summary>
         /// <param name="mainGoal">The main goal structure that the agent needs to complete.</param>
-        public DesireSet(IGoalStructure<TBeliefSet> mainGoal)
-            => _mainGoal = mainGoal;
+        /// <param name="sideGoals">The side goal structures that could be activated during the agent playthrough.</param>
+        public DesireSet(
+            IGoalStructure<TBeliefSet> mainGoal,
+            params (IGoalStructure<TBeliefSet> goalStructure, System.Func<TBeliefSet, bool> guard)[] sideGoals
+        )
+        {
+            _mainGoal = mainGoal;
+            _goalStructureStack = new(sideGoals);
+
+            // Push the main goal structure on the stack.
+            _goalStructureStack.Activate(new((_mainGoal, _ => false), _goalStructureStack));
+        }
+
+        /// <summary>
+        /// Pushes side goal structures on the stack if their guard is fulfilled.
+        /// </summary>
+        /// <param name="beliefSet">The belief set to check the guards of the goal structures with.</param>
+        private void ActivateRelevantGoalStructures(TBeliefSet beliefSet)
+        {
+            foreach (var goalStructureStackItem in _goalStructureStack.ActivatableStackItems)
+            {
+                System.Func<TBeliefSet, bool> guard = goalStructureStackItem.Data.guard;
+
+                if (!guard(beliefSet)) continue;
+
+                _goalStructureStack.Activate(goalStructureStackItem);
+            }
```

Comment:
This can also be done with a LINQ statement:
```suggestion
			// Filter all the items by their guard.
            IEnumerable itemsToActivate = _goalStructureStack.ActivatableStackItems
                .Where(item => item.Data.guard(beliefSet))
            
            // Activate the stacks of the interrupts.
            foreach (var item in itemsToActivate)
            {
                _goalStructureStack.Activate(item);
            }
```

I think it is nicer and better understandable this way.

Reply:
Agreed. Nice solution